tests/mqtt_mock_server_handler.h

```diff
@@ -407,3 +467,152 @@ static void s_mqtt_mock_server_wait_for_pubacks(struct aws_channel_handler *hand
+    size_t length = aws_array_list_length(&received_messages);
+    for (size_t i = testing_handler->decoded_index; i < length; i++) {
+        struct aws_byte_buf received_message = {0};
+        ASSERT_SUCCESS(aws_array_list_get_at(&received_messages, &received_message, i));
+        struct aws_byte_cursor message_cur = aws_byte_cursor_from_buf(&received_message);
+
+        struct mqtt_decoded_packet packet;
+        ASSERT_SUCCESS(s_mqtt_decode_packet_init(&packet, testing_handler->handler.alloc));
+        packet.type = aws_mqtt_get_packet_type(message_cur.ptr);
+
+        struct aws_mqtt_packet_connect connect_packet;
+        struct aws_mqtt_packet_subscribe subscribe_packet;
+        struct aws_mqtt_packet_unsubscribe unsubscribe_packet;
+        struct aws_mqtt_packet_publish publish_packet;
+        struct aws_mqtt_packet_ack puback;
+
+        switch (packet.type) {
+            case AWS_MQTT_PACKET_CONNECT:
+                ASSERT_SUCCESS(aws_mqtt_packet_connect_decode(&message_cur, &connect_packet));
+                packet.clean_session = connect_packet.clean_session;
+                packet.has_will = connect_packet.has_will;
+                packet.will_retain = connect_packet.will_retain;
+                packet.has_password = connect_packet.has_password;
+                packet.has_username = connect_packet.has_username;
+                packet.keep_alive_timeout = connect_packet.keep_alive_timeout;
+                packet.will_qos = connect_packet.will_qos;
+
+                ASSERT_SUCCESS(s_move_cursor_to_buf_dynamic(
+                    &packet.buffer, &connect_packet.client_identifier, &packet.client_identifier));
+                if (packet.has_will) {
+                    ASSERT_SUCCESS(
+                        s_move_cursor_to_buf_dynamic(&packet.buffer, &connect_packet.will_topic, &packet.will_topic));
+                    ASSERT_SUCCESS(s_move_cursor_to_buf_dynamic(
+                        &packet.buffer, &connect_packet.will_message, &packet.will_message));
+                }
+                if (packet.has_username) {
+                    ASSERT_SUCCESS(
+                        s_move_cursor_to_buf_dynamic(&packet.buffer, &connect_packet.username, &packet.username));
+                }
+                if (packet.has_password) {
+                    ASSERT_SUCCESS(
+                        s_move_cursor_to_buf_dynamic(&packet.buffer, &connect_packet.password, &packet.password));
+                }
+                break;
+
+            case AWS_MQTT_PACKET_SUBSCRIBE:
+                ASSERT_SUCCESS(aws_mqtt_packet_subscribe_init(&subscribe_packet, testing_handler->handler.alloc, 0));
+                ASSERT_SUCCESS(aws_mqtt_packet_subscribe_decode(&message_cur, &subscribe_packet));
+                packet.packet_identifier = subscribe_packet.packet_identifier;
+                /* copy the array one by one for simplicity */
+                while (aws_array_list_length(&subscribe_packet.topic_filters)) {
```

Comment:
Why this instead of a non-destructive for-loop?

Reply:
Good call! for-loop will be much better and efficient here.